### PR TITLE
Update ParaView to v5.4.0

### PR DESCRIPTION
**Pull request by quantumsteve**
Description of work.

This updates Mantid to use [ParaView 5.4.0](https://blog.kitware.com/paraview-5-4-0-release-notes/). Users will notice the updated color legend that chooses nicer increments that can be represented with fewer digits. For the VSI, we remove our custom `SaveScreenshotReaction` in favor of the refactored `pqSaveScreenshotReaction` class. Since "AllViews" is no longer selected by default, our custom changes are no longer necessary.

I wasn't able to remove the `-Wpedantic` warning from `vtktypemacro`. Given the potential dangers from extra semicolons, I removed this flag within the Vates subdirectory.

I created a [separate issue](https://github.com/mantidproject/mantid/issues/19946) to address complications with ParaView 5.4 and older system versions of jsoncpp.

**To test:**

1. View the updated [build instructions](https://www.mantidproject.org/Building_VATES).
2. Build Mantid with -DMAKE_VATES=ON and -DParaView_DIR=\<your ParaView build directory\>.
3. Try viewing a MDHistoWorkspace and a MDEventWorkspace in the VSI.
4. Save a screenshot or two.

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by quantumsteve** [comment]
It's now building successfully on all platforms :tada:

**Review by AndreiSavici — APPROVED** [review]
It seems to work on build servers. Code review is OK. I have problems building it on Ubuntu 16.10 (tbb library). Time for me to upgrade

**Review by AntonPiccardoSelg — APPROVED** [review]
Changes look fine. MDHisto and MDEvent operations seem to be fine. Taking a screenshot works. Tested on Ubuntu `16.04`

**Comment by martyngigg** [comment]
@quantumsteve I built ParaView 5.4.0 with the buildscript on the `v5.4.0` branch and built mantid against this. I was able to visualise an `MDHistoWorkspace` but I then loaded `SEQ_11499_md_enp.nxs` from the system test data and I now get a crash when visualising this. The console contains some errors messages:

```
ERROR: In /media/data1/source/paraview/ParaView-v5.4.0-source/VTK/Filters/Core/vtkMaskPoints.cxx, line 320
vtkMaskPoints (0x7edd180): No points to mask
```

I tried this with the current nightly and I actually get this same error but in the usual error window and Mantid does not crash, it just doesn't display anything.

**Comment by quantumsteve** [comment]
@martyngigg I was able to reproduce the issue and _think_ it's an internal ParaView issue. Specifically, it looks like `vtkPVProgressHandler::GetLastMessage()` may return `NULL` and passing `NULL` to the `QString` constructor in `pqProgressManager::onMessage` is causing the segfault.

Does this match what you are seeing?

**Comment by quantumsteve** [comment]
Commenting out everything in `pqProgressManager::onMessage` so that it is a noop appears to "fix" the segfault. Otherwise I seem to get some sort of infinite recursion and eventually a corrupt pointer.

**Comment by martyngigg** [comment]
It looks to me as if `pqProgressManager::onMessage` calls `vtkOutputWindow::DisplayErrorText` which then calls `this->InvokeEvent(vtkCommand::MessageEvent, (void*)txt);` again that causes it to end up in `pqProgressManager::onMessage` again.

It's not entirely clear what starts this whole chain though...

**Review by martyngigg — CHANGES_REQUESTED** [review]
I'm just marking this as not approved so that it doesn't get merged accidentally.

**Comment by quantumsteve** [comment]
The last commit forces at least one box to be selected. This avoids the segfault @martyngigg found, though doesn't fix the lower-level issue.

**Comment by quantumsteve** [comment]
I suspect it is related to this commit and turning on `VTK_LEGACY_REMOVE`
https://github.com/Kitware/ParaView/commit/2e16d4a2c27bd350b6337d8ef5e1a2e5c9b9bb77

**Comment by quantumsteve** [comment]
Here is the `pqOutputWidget` that replaces `pqOutputWindow`. 
https://github.com/Kitware/ParaView/commit/bb4df3609eaa8dad9aa0a16310d4ffd5b90de02b

I also tried turning off `VTK_LEGACY_REMOVE` and get a segfault in `MdViewerWidget::renderWorkspace`. It might be easier to migrate to the new interface.

**Comment by martyngigg** [comment]
@quantumsteve Migrating sounds like a better long term plan.

**Comment by quantumsteve** [comment]
@martyngigg The last commit should be enough to match the behavior from v5.3.0. Switching to QMainWindow to take advantage of QDockWidget looks like a significant refactoring that should be done as a separate pull request.

File: Vates/VatesSimpleGui/ViewWidgets/inc/MantidVatesSimpleGuiViewWidgets/MdViewerWidget.h [hunk 1]
```diff
@@ -1,6 +1,9 @@
 #ifndef MDVIEWERWIDGET_H_
 #define MDVIEWERWIDGET_H_
 
+#ifdef ERROR
+#undef ERROR
+#endif
```
Comment (quantumsteve) [inline Vates/VatesSimpleGui/ViewWidgets/inc/MantidVatesSimpleGuiViewWidgets/MdViewerWidget.h]: @martyngigg Any idea where the `ERROR` macro is coming from? Without this, windows won't build.

**Comment by martyngigg** [comment]
@quantumsteve It seems to be coming from the `wingdi.h` header. It contains a `#define ERROR 0` line - 
 see https://github.com/baysao/google-glog/issues/33 for a similar error.

`wingdi.h` is included by the main `Windows.h` header which I assume is being included by Qt somewhere along the stack. The issue linked suggests using a project-wide `NOGDI` define could help. I assume this would be okay as Qt is the one doing to graphics handling and not us. I'm just compiling it on Windows to check.

**Comment by martyngigg** [comment]
I tried compiling on Windows with `NOGDI` defined but it didn't help. I think we may have to live with this "workaround for now".

I can also confirm that the VSI no longer crashes when the "No points to mask" is encountered. However, now the output widget seems to be stuck where it is created. See screenshot. I can interact with it with the exception of not being able to drag the window around.

**Comment by quantumsteve** [comment]
@martyngigg I tried removing `QDockWindow` and having it mostly working on OSX. The output window is shown, but is in the background. 

Please let me know if this addresses the larger issues you encountered on Ubuntu.

**Comment by martyngigg** [comment]
@quantumsteve This seems to address the issue perfectly on Ubuntu. The window also appears on top of the VSI as I would expect.

**Comment by quantumsteve** [comment]
@martyngigg Having the floating output window in the background on OSX is a minor issue and probably related to our use of setModal(true) on OS X. 

Is this good enough to merge?

**Review by rosswhitfield — APPROVED** [review]
Works on Arch Linux :+1:

**Review by martyngigg — APPROVED** [review]

**Pull request merged**